<h1>App Physics 155 (lab)</h1>
<h2>Feb. 7, 2018</h2><br />
Kenneth V. Domingo<br />
2015-03116

In [18]:
import numpy as np
from decimal import Decimal as dec
from decimal import getcontext

<b>Exercise 4.2: Quadratic equations</b><br /><br />
a) Write a program that takes as input three numbers, $a$, $b$, and $c$, and prints out
the two solutions to the quadratic equation $ax^2 + bx + c = 0$ using the standard
formula

\begin{equation}
    x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}
\end{equation}

Use your program to compute the solutions of $0.001x^2 + 1000x + 0.001 = 0$.

In [19]:
def quadratic(a,b,c):
    x1 = (-b + np.sqrt(b**2 - 4*a*c))/(2*a)
    x2 = (-b - np.sqrt(b**2 - 4*a*c))/(2*a)
    print("x1 = ", x1)
    print("x2 = ", x2)
    print("Solution 1 = ", a*x1**2 + b*x1 + c)
    print("Solution 2 = ", a*x2**2 + b*x2 + c)

In [20]:
quadratic(.001, 1000, .001)

x1 =  -9.99989424599e-07
x2 =  -999999.999999
Solution 1 =  1.05754016655e-08
Solution 2 =  7.24792480469e-08


b) There is another way to write the solutions to a quadratic equation. Multiplying
top and bottom of the solution above by $−b \mp \sqrt{b^2 − 4ac}$, show that the solutions
can also be written as

\begin{equation}
    x = \frac{2c}{-b \mp \sqrt{b^2 - 4ac}}
\end{equation}

Add further lines to your program to print these values in addition to the earlier
ones and again use the program to solve $0.001x^2 + 1000x + 0.001 = 0$. What do
you see? How do you explain it?

In [ ]:
def quadratic_acc(a,b,c):
    x1 = (2*c)/(-b - np.sqrt(b**2 - 4*a*c))
    x2 = (2*c)/(-b + np.sqrt(b**2 - 4*a*c))
    print("x1 = ", x1)
    print("x2 = ", x2)
    print("Solution 1 = ", a*x1**2 + b*x1 + c)
    print("Solution 2 = ", a*x2**2 + b*x2 + c)

In [ ]:
quadratic_acc(.001,1000,.001)

(c) Using what you have learned, write a new program that calculates both roots of
a quadratic equation accurately in all cases.

In [49]:
getcontext().prec = 16
def quadratic_prec1(a,b,c):
    x1 = (dec(-b) + np.sqrt(dec(b)**dec(2) - dec(4)*dec(a)*dec(c)))/(dec(2)*dec(a))
    x2 = (dec(-b) - np.sqrt(dec(b)**dec(2) - dec(4)*dec(a)*dec(c)))/(dec(2)*dec(a))
    sol1 = dec(a)*dec(x1)**dec(2) + dec(b)*dec(x1) + dec(c)
    sol2 = dec(a)*dec(x2)**dec(2) + dec(b)*dec(x2) + dec(c)
    print("By equation (1): ")
    print("x1 = ", x1)
    print("x2 = ", x2)
    print("Solution 1 = ", sol1)
    print("Solution 2 = ", sol2)
    
def quadratic_prec2(a,b,c):
    x1 = (dec(2)*dec(c))/(dec(-b) - np.sqrt(dec(b)**dec(2) - dec(4)*dec(a)*dec(c)))
    x2 = (dec(2)*dec(c))/(dec(-b) + np.sqrt(dec(b)**dec(2) - dec(4)*dec(a)*dec(c)))
    sol1 = dec(a)*dec(x1)**dec(2) + dec(b)*dec(x1) + dec(c)
    sol2 = dec(a)*dec(x2)**dec(2) + dec(b)*dec(x2) + dec(c)
    print("\nBy equation (2): ")
    print("x1 = ", x1)
    print("x2 = ", x2)
    print("Solution 1 = ", sol1)
    print("Solution 2 = ", sol2)

In [50]:
quadratic_prec1(.001,1000,.001)
quadratic_prec2(.001,1000,.001)

By equation (1): 
x1 =  -0.000001
x2 =  -999999.999999
Solution 1 =  1.000020816681712E-15
Solution 2 =  2.081668171172169E-20

By equation (2): 
x1 =  -0.000001000000000001000
x2 =  -1000000.00000
Solution 1 =  2.081668171172169E-20
Solution 2 =  0.001000000000000000


Equation (2) yields more accurate roots for the quadratic equation, where

\begin{equation}
    x_{+} \approx -0.000001
\end{equation}
\begin{equation}
    x_{-} \approx -1000000
\end{equation}

Plugging these into the original quadratic equation, we get

\begin{equation}
    x_{-}: 2.08 \times 10^{-20} \approx 0
\end{equation}
\begin{equation}
    x_{+}: 0.001 \sim 0
\end{equation}